### ISSUE TITLE
CPU triangular solve is slow

### ISSUE THREAD

**Issue by yaroslavvb** [pytorch/pytorch#12683]
It would be nice to have a GPU version of triangular solve. It's needed for preconditioned stochastic gradient optimizer ([paper](https://ieeexplore.ieee.org/document/7875097/)), and it's possibly the reason why TF implementation runs 3x faster than PyTorch version

Here's a microbenchmark in milliseconds of 8000x8000 triangular solve:
https://github.com/yaroslavvb/newton/blob/master/benchmark_triangular.py

```
numpy: Times: min: 127.86, median: 128.12, mean: 128.31
tensorflow: Times: min: 1.20, median: 1.20, mean: 1.20
pytorch: Times: min: 162.65, median: 162.65, mean: 162.65
```

cc @VitalyFedyunin @jianyuh @nikitaved @pearu @mruberry @heitorschueroff @walterddr @IvanYashchuk @xwang233 @ngimel

**Comment by yaroslavvb** [comment]
I just realized docs are out of date and there's indeed a gpu implementation

**Comment by vishwakftw** [comment]
@yaroslavvb I believe the CUDA threads are not synchronized in your benchmark, could you please check?

**Comment by yaroslavvb** [comment]
I've updated numbers with sync points. TensorFlow CPU version probably benefits from being compiled specifically for p3 CPU architecture (using DLAMI)

```
Benchmarking n=10000
numpy
Times: min: 202.54, median: 203.09, mean: 203.31
Pytorch GPU
Times: min: 6.46, median: 6.49, mean: 6.52
Pytorch CPU
Times: min: 254.73, median: 260.98, mean: 261.95
TF GPU
Times: min: 1.53, median: 1.60, mean: 1.60
TF CPU
Times: min: 52.40, median: 54.21, mean: 53.87
```

**Comment by lezcano** [comment]
It looks like things have improved in the last 5 years. With the following script
```python
import torch
import numpy as np
import scipy
from torch.utils.benchmark import Timer, Compare


bench_results = []
n = 8000

A = scipy.randn(n, n)
A = A @ A.T + np.eye(n)
A = scipy.linalg.cholesky(A, lower=True)  # upper diagonal matrix
b = scipy.randn(n)
timer = Timer(stmt="scipy.linalg.solve_triangular(A, b, lower=True)",
              globals=locals(),
              label="Solve Triangular",
              description="cpu",
              sub_label="scipy")
bench_results.append(timer.blocked_autorange())

A = torch.as_tensor(A)
b = torch.as_tensor(b).unsqueeze(-1)
timer = Timer(stmt="torch.linalg.solve_triangular(A, b, upper=False)",
              globals=locals(),
              label="Solve Triangular",
              description="cpu",
              sub_label="torch")
bench_results.append(timer.blocked_autorange())
compare = Compare(bench_results)
compare.trim_significant_figures()
compare.print()
```

I'm getting:
```python
[- Solve Triangular ]
             |   cpu 
1 threads: ----------
      scipy  |  50.13
      torch  |   9.57

Times are in milliseconds (ms).
```
